resources/aws_ec2.go

```diff
@@ -0,0 +1,535 @@
+	client *ec2.EC2 // client session for AWS API calls
+	wg     *sync.WaitGroup
+}
+
+// Default returns some sensible defaults for this resource.
+func (obj *AwsEc2Res) Default() Res {
+	return &AwsEc2Res{
+		BaseRes: BaseRes{
+			MetaParams: DefaultMetaParams, // force a default
+		},
+	}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *AwsEc2Res) Validate() error {
+	if obj.State != "running" && obj.State != "stopped" && obj.State != "terminated" {
+		return fmt.Errorf("state must be 'running', 'stopped' or 'terminated'")
+	}
+
+	// compare obj.Region to the list of available AWS endpoints.
+	validRegion := false
+	for _, region := range AwsRegions {
+		if obj.Region == region {
+			validRegion = true
+		}
+	}
+	if !validRegion {
+		return fmt.Errorf("region must be a valid AWS endpoint")
+	}
+
+	// check the instance type
+	// there is currently no api call to enumerate available instance types
+	if obj.Type == "" {
+		return fmt.Errorf("no instance type specified")
+	}
+
+	// check imageId against a list of available images
+	sess, err := session.NewSession(&aws.Config{
+		Region: aws.String(obj.Region),
+	})
+	if err != nil {
+		return errwrap.Wrapf(err, "error creating session")
+	}
+	client := ec2.New(sess)
+
+	imagesInput := &ec2.DescribeImagesInput{}
+	images, err := client.DescribeImages(imagesInput)
+	if err != nil {
+		return errwrap.Wrapf(err, "error describing images")
+	}
+	validImage := false
+	for _, image := range images.Images {
+		if obj.ImageID == *image.ImageId {
+			validImage = true
+		}
+	}
+	if !validImage {
+		return fmt.Errorf("imageid must be a valid ami available in the specified region")
+	}
+
+	return obj.BaseRes.Validate()
+}
+
+// Init initializes the resource.
+func (obj *AwsEc2Res) Init() error {
+	// create a client session for the AWS API
+	sess, err := session.NewSession(&aws.Config{
+		Region: aws.String(obj.Region),
+	})
+	if err != nil {
+		return errwrap.Wrapf(err, "error creating session")
+	}
+	obj.client = ec2.New(sess)
+
+	obj.wg = &sync.WaitGroup{}
+
+	return obj.BaseRes.Init() // call base init, b/c we're overriding
+}
+
+// Watch is the primary listener for this resource and it outputs events.
+func (obj *AwsEc2Res) Watch() error {
+	send := false
+	var exit *error
+	if err := obj.Running(); err != nil {
+		return err
+	}
+	type chanStruct struct {
+		str string
+		err error
+	}
+	awsChan := make(chan *chanStruct)
+	closeChan := make(chan struct{})
+	ctx, cancel := context.WithCancel(context.TODO())
+	obj.wg.Add(1)
+	go func() {
+		defer obj.wg.Done()
+		select {
+		case <-closeChan:
+			cancel()
+		}
+	}()
+	obj.wg.Add(1)
+	defer obj.wg.Wait()
+	go func() {
+		defer obj.wg.Done()
+		defer close(awsChan)
+		for {
+			diInput := &ec2.DescribeInstancesInput{
+				Filters: []*ec2.Filter{
+					{
+						Name:   aws.String("tag:Name"),
+						Values: []*string{aws.String(obj.prependName())},
+					},
+					{
+						Name: aws.String("instance-state-name"),
+						Values: []*string{
```

Comment:
you had a clever idea about looking at the requested state and adding the opposite string here, are we still doing that?

Reply:
that happens lower down (stoppedInput/runningInput), this request gets us the instance ids we need for the WaitUntil requests.